crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,288 @@
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+
+        let manifest_path = self
+            .cache
+            .manifest_for_reference(&reference.to_string())
+            .await?;
+        let config_path = self
+            .cache
+            .config_for_reference(&reference.to_string())
+            .await?;
+        if !manifest_path.exists() || !config_path.exists() {
+            self.pull(&reference.to_string()).await?;
+        }
+
+        let config = serde_json::from_slice(&Self::data(&config_path).await?)?;
+        let manifest = serde_json::from_slice(&Self::data(&manifest_path).await?)?;
+
+        Ok(SpinOciDescriptor {
+            reference: reference.to_string(),
+            manifest,
+            config,
+        })
+    }
+
+    /// Create a new wasm layer based on a file.
+    pub async fn wasm_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading wasm module from {:?}", file);
+        Ok(ImageLayer::new(
+            Self::data(file).await.context("cannot read wasm module")?,
+            WASM_LAYER_MEDIA_TYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Create a new data layer based on a file.
+    pub async fn data_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading data file from {:?}", file);
+        Ok(ImageLayer::new(
+            Self::data(&file).await?,
+            DATA_MEDIATYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Construct the registry authentication based on the reference.
+    pub fn auth(reference: &Reference) -> Result<RegistryAuth> {
+        let server = reference
+            .resolve_registry()
+            .strip_suffix("/")
+            .unwrap_or_else(|| reference.resolve_registry());
+
+        match docker_credential::get_credential(server) {
+            Err(CredentialRetrievalError::ConfigNotFound) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::NoCredentialConfigured) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::ConfigReadError) => Ok(RegistryAuth::Anonymous),
+            Err(e) => bail!("Error handling docker configuration file: {:?}", e),
+
+            Ok(DockerCredential::UsernamePassword(username, password)) => {
+                tracing::trace!("Found docker credentials");
+                Ok(RegistryAuth::Basic(username, password))
+            }
+            Ok(DockerCredential::IdentityToken(_)) => {
+                tracing::warn!("Cannot use contents of docker config, identity token not supported. Using anonymous auth");
+                Ok(RegistryAuth::Anonymous)
+            }
+        }
+    }
+
+    /// Build the OCI client configuration given the insecure option.
+    pub fn build_config(insecure: bool) -> ClientConfig {
+        let protocol = if insecure {
+            oci_distribution::client::ClientProtocol::Http
```

Comment:
The help text for --insecure shows "Ignore server certificate errors". It does not specify that it switches the protocol to HTTP.

We should either put this behind a different flag, or update the help text.